src/skdh/context/motion_detect.py

```diff
@@ -38,42 +58,301 @@ def __init__(self, filter_cutoff=6, cov_threshold=0.022558169349678834):
+    def predict(self, time, accel, fs=None, **kwargs):
+        """
+        predict(time, accel, fs)
+
+        Function to detect periods of motion from an accelerometer signal.
+
+        Parameters
+        ----------
+        time : numpy.ndarray
+            (N, ) array of unix timestamps, in seconds.
+        accel : numpy.ndarray
+            (N, 3) array of acceleration, in units of 'g', collected at 20hz.
+        fs : float
+            Sampling rate. Default None. If not provided, will be inferred.
+
+        Returns
+        -------
+        results : dict
+            Results dictionary with motion predictions (1hz boolean; true indicates motion) and associated timestamps.
+
+        """
+        # check input requirements are met
+        time, accel, fs = self._check_input(time, accel, fs)
+
+        # Branch 1: Rolling CoV thresholding
+        # 1. Vector Magnitude
+        vmag = np.linalg.norm(accel, axis=1)
+
+        # 2. Low-pass filter the accelerometer vector magnitude signal to remove high frequency components
+        sos_lp = butter(
+            N=self.lp_filter_order,
+            Wn=2 * self.lp_filter_cutoff / fs,
+            btype="lowpass",
+            output="sos",
+        )
+        vmag_lp = sosfiltfilt(sos_lp, vmag, axis=0)
+
+        # 3. High-pass filter the accelerometer vector magnitude signal to remove DC component
+        sos_hp = butter(
+            N=self.hp_filter_order,
+            Wn=2 * self.hp_filter_cutoff / fs,
+            btype="highpass",
+            output="sos",
+        )
+        vmag_lp_hp = sosfiltfilt(sos_hp, vmag_lp, axis=0)
+
+        # 4. Calculate the 1s rolling coefficient of variation
+        rolling_std, rolling_mean = moving_sd(a=vmag_lp_hp, w_len=int(fs), skip=1)
+        rolling_cov = rolling_std / (np.abs(rolling_mean) + 0.01)
+
+        # 5. Detect CoV values greater than movement threshold
+        movement = rolling_cov >= self.cov_threshold
+
+        # 6. Get non-overlapping mean of predictions. If average of predictions in window is >=0.5 (at least half of
+        # the CoVs are greater than the threshold), then movement is true.
+        windowed_movement = moving_mean(a=movement, w_len=int(fs), skip=int(fs))
+        cov_predictions = windowed_movement >= 0.5
+
+        # Branch 2: Max standard deviation thresholding
+        # 1. Moving SD with no overlap
+        sd, _ = moving_sd(a=accel, w_len=int(fs), skip=int(fs), axis=0)
+
+        # 2. Max in each second
+        maxsd = np.max(sd, axis=1)
+
+        # 3. Threshold maxsd
+        maxsd_predictions = maxsd >= self.sd_threshold
+
+        # Final step: OR of branch 1 and 2
+        predictions = cov_predictions | maxsd_predictions[0:len(cov_predictions)]
+
+        # Sample time array to 1s resolution
+        time_1s = time[0 :: int(fs)].copy()[0:len(predictions)]
+
+        # compile results
+        results = {"movement_detected": predictions, "movement_time": time_1s}
```

Comment:
same as previous method - suggest reformating these results to start/end of movement periods, `results_to_kwargs` should be set to `True` if these are used downstream as well

Reply:
Changed